{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97b69a85",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.messages import SystemMessage, HumanMessage\n",
    "from langgraph.graph import StateGraph, START, END\n",
    "from langgraph.constants import Send\n",
    "from typing_extensions import List, Dict, TypedDict, Annotated\n",
    "from pydantic import BaseModel\n",
    "import json\n",
    "from operator import add\n",
    "from rag_eval.prompt_manager import PromptManager\n",
    "from rag_eval.utils import create_path_if_not_exists\n",
    "from rag_eval import ChunkDataHandler\n",
    "import os\n",
    "from fcgb.cfg.precompiled import get_llm, get_checkpointer\n",
    "\n",
    "import asyncio\n",
    "from tqdm.notebook import tqdm as notebook_tqdm\n",
    "import uuid\n",
    "\n",
    "import sys\n",
    "sys.path.append('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3ae12eea",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HydeAnswerModel(BaseModel):\n",
    "    answers: List[str]\n",
    "\n",
    "class HyDEset(TypedDict):\n",
    "    query: str\n",
    "    hyde_queries: List[str]\n",
    "\n",
    "class HyDEState(BaseModel):\n",
    "    title: str\n",
    "    summary: str = ''\n",
    "    query: str = ''\n",
    "\n",
    "class HyDEGenState(BaseModel):\n",
    "    doc_name: str\n",
    "    title: str = ''\n",
    "    summary: str = ''\n",
    "    queries: List[str] = []\n",
    "    hyde_sets: Annotated[List[HyDEset], add] = []\n",
    "\n",
    "\n",
    "class HyDEQueriesGenerator(ChunkDataHandler):\n",
    "    def __init__(\n",
    "            self,\n",
    "            llm,\n",
    "            data_path: str,\n",
    "            prompts_config: Dict,\n",
    "            memory=None,\n",
    "            hyde_queries: int = 5,\n",
    "            prompt_manager_spec: Dict = {}\n",
    "    ):\n",
    "        \"\"\"\n",
    "        Initializes the HyDEQueriesGenerator with the necessary parameters.\n",
    "        Args:\n",
    "            llm: The language model to use for generating queries.\n",
    "            hyde_path (str): Output path for HyDE queries.\n",
    "            data_path: Main path of chunk dataset.\n",
    "            prompts_config (Dict): Configuration for prompts.\n",
    "                Configuration should include:\n",
    "                'path': main path where prompts are stored,\n",
    "                'hyde_queries': prompt for generating HyDE queries.\n",
    "            memory: Optional memory object for storing state.\n",
    "            hyde_queries: Number of queries to generate (default is 5).\n",
    "            prompt_manager_spec: Specifications for the prompt manager.\n",
    "        \"\"\"\n",
    "        super().__init__(output_path=data_path)\n",
    "\n",
    "        self.llm = llm\n",
    "        self.hyde_queries = hyde_queries\n",
    "        self.memory = memory\n",
    "        self.prompts_config = prompts_config\n",
    "        self.prompt_manager_spec = prompt_manager_spec\n",
    "\n",
    "        self.hyde_path = os.path.join(data_path, 'hyde_queries')\n",
    "        create_path_if_not_exists(self.hyde_path)\n",
    "\n",
    "        self._set_prompts()\n",
    "        self.build_graph()\n",
    "\n",
    "    def _set_prompts(self):\n",
    "        \"\"\"\n",
    "        Initializes the prompt manager and retrieves the prompts for generating queries.\n",
    "        \"\"\"\n",
    "        prompt_manager = PromptManager(\n",
    "            version_config=self.prompt_manager_spec,\n",
    "            path=self.prompts_config['path']\n",
    "        )\n",
    "\n",
    "        self.prompts = prompt_manager.get_prompts([self.prompts_config[name] for name in ['hyde_queries']])\n",
    "\n",
    "    def load_metadata_node(self, state: HyDEGenState):\n",
    "        \"\"\"\n",
    "        Loads metadata for the document specified in the state.\n",
    "        \"\"\"\n",
    "        metadata = self.get_metadata_file(state.doc_name)\n",
    "\n",
    "        return {\n",
    "            'title': metadata.get('title', ''),\n",
    "            'summary': metadata.get('summary', ''),\n",
    "            'queries': metadata.get('queries', []),\n",
    "        }\n",
    "    \n",
    "    def queries_routing(self, state: HyDEGenState):\n",
    "        \"\"\"\n",
    "        Distributes the queries to the HyDE generation nodes.\n",
    "        \"\"\"\n",
    "        return [Send('hyde_gen', HyDEState(\n",
    "                    title=state.title,\n",
    "                    summary=state.summary,\n",
    "                    query=query)\n",
    "                    ) for query in state.queries]\n",
    "    \n",
    "    def hyde_gen(self, state: HyDEState):\n",
    "        \"\"\"\n",
    "        Generates HyDE queries in parallel for every base query.\n",
    "        \"\"\"\n",
    "        hyde_prompt = self.prompts['hyde_queries'].format(\n",
    "            title=state.title,\n",
    "            summary=state.summary,\n",
    "            query=state.query,\n",
    "            hyde_queries_num=self.hyde_queries\n",
    "        )\n",
    "\n",
    "        hyde_queries = self.llm.with_structured_output(HydeAnswerModel).invoke(hyde_prompt).answers\n",
    "\n",
    "        hyde_output = HyDEset(\n",
    "            query=state.query,\n",
    "            hyde_queries=hyde_queries\n",
    "        )\n",
    "        return {'hyde_sets': [hyde_output]}\n",
    "    \n",
    "    def save_hyde_queries(self, state: HyDEGenState):\n",
    "        \"\"\"\n",
    "        Saves the generated HyDE queries to a JSON file.\n",
    "        \"\"\"\n",
    "        \n",
    "        hyde_file_path = os.path.join(self.hyde_path, f\"{state.doc_name}.json\")\n",
    "        with open(hyde_file_path, 'w') as f:\n",
    "            json.dump({'queries_set': state.hyde_sets}, f, indent=4)\n",
    "\n",
    "        return {}\n",
    "    \n",
    "    def build_graph(self):\n",
    "        \"\"\"\n",
    "        Builds the state graph for the HyDE queries generation workflow.\n",
    "        \"\"\"\n",
    "        workflow = StateGraph(HyDEGenState)\n",
    "        workflow.add_node('load_metadata', self.load_metadata_node)\n",
    "        workflow.add_node('hyde_gen', self.hyde_gen)\n",
    "        workflow.add_node('save_hyde_queries', self.save_hyde_queries)\n",
    "\n",
    "        workflow.add_edge(START, 'load_metadata')\n",
    "        workflow.add_conditional_edges('load_metadata', self.queries_routing, 'hyde_gen')\n",
    "        workflow.add_edge('hyde_gen', 'save_hyde_queries')\n",
    "        workflow.add_edge('save_hyde_queries', END)\n",
    "\n",
    "        self.graph = workflow.compile(checkpointer=self.memory)\n",
    "\n",
    "    def run(self, doc_name: str, thread_id: str = None):\n",
    "        \"\"\"\n",
    "        Runs the HyDE queries generation workflow for a given document name.\n",
    "        Args:\n",
    "            doc_name (str): The name of the document for which to generate HyDE queries.\n",
    "            thread_id (str, optional): An identifier for the thread, if applicable.\n",
    "        Returns:\n",
    "            HyDEGenState: The final state of the workflow containing the generated HyDE queries.\n",
    "        \"\"\"\n",
    "        config = {'configurable': {'thread_id': thread_id}} if thread_id else None\n",
    "        return self.graph.invoke({'doc_name': doc_name}, config=config)\n",
    "    \n",
    "    async def add_hyde_queries(self, target_docs: int, concurrent_runs=5):\n",
    "\n",
    "        docs_to_process = self._docs_without_hyde()\n",
    "        current_docs_processed = len(self._docs_with_hyde())\n",
    "        docs_needed = max(target_docs - current_docs_processed, 0)\n",
    "\n",
    "        docs_to_process = docs_to_process[:docs_needed]\n",
    "\n",
    "        process_pbar = notebook_tqdm(\n",
    "            total=len(docs_to_process),\n",
    "            desc='HyDE queries generation',\n",
    "            unit='doc',\n",
    "            postfix={'Target docs': target_docs, 'Processed docs': current_docs_processed}\n",
    "        )\n",
    "\n",
    "        async def worker(queue):\n",
    "            while not queue.empty():\n",
    "                doc_name = await queue.get()\n",
    "                try:\n",
    "                    thread_id = uuid.uuid4().hex\n",
    "                    state = self.run(doc_name, thread_id=thread_id)\n",
    "                    process_pbar.update(1)\n",
    "                    process_pbar.set_postfix({'Target docs': target_docs, 'Processed docs': len(self._docs_with_hyde())})\n",
    "                except Exception as e:\n",
    "                    print(f\"Error processing {doc_name}: {e}\")\n",
    "                finally:\n",
    "                    queue.task_done()\n",
    "\n",
    "        queue = asyncio.Queue()\n",
    "        for doc_name in docs_to_process:\n",
    "            await queue.put(doc_name)\n",
    "\n",
    "        workers = [asyncio.create_task(worker(queue)) for _ in range(concurrent_runs)]\n",
    "        await queue.join()\n",
    "\n",
    "        await asyncio.gather(*workers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7b06f4ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyde_gen = HyDEQueriesGenerator(\n",
    "    llm=get_llm(llm_model='google'),\n",
    "    data_path='../data',\n",
    "    prompts_config={\n",
    "        'path': None,\n",
    "        'hyde_queries': 'hyde_queries'\n",
    "    },\n",
    "    memory=get_checkpointer(checkpointer_mode='local'),\n",
    "    hyde_queries=5,\n",
    "    prompt_manager_spec={}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5cf41b35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All docs: 115\n",
      "Docs to evaluate: 15 - 13.04% of all docs\n",
      "Docs evaluated: 100 - 86.96% of all docs\n",
      "Docs with HyDE queries: 100 - 100.00% of evaluated docs\n"
     ]
    }
   ],
   "source": [
    "hyde_gen.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b49e9fa",
   "metadata": {},
   "source": [
    "### database extension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6a326ef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ed5c9cb4784743cba2468968258d019e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HyDE queries generation:   0%|          | 0/80 [00:00<?, ?doc/s, Processed docs=20, Target docs=100]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "await hyde_gen.add_hyde_queries(target_docs=100, concurrent_runs=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b6fd9f5",
   "metadata": {},
   "source": [
    "### single run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "134b9250",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'doc_name': '1705.03998v1',\n",
       " 'title': 'Mining Functional Modules by Multiview-NMF of Phenome-Genome Association',\n",
       " 'summary': 'Background: Mining gene modules from genomic data is an important step to\\ndetect gene members of pathways or other relations such as protein-protein\\ninteractions. In this work, we explore the plausibility of detecting gene\\nmodules by factorizing gene-phenotype associations from a phenotype ontology\\nrather than the conventionally used gene expression data. In particular, the\\nhierarchical structure of ontology has not been sufficiently utilized in\\nclustering genes while functionally related genes are consistently associated\\nwith phenotypes on the same path in the phenotype ontology. Results: We propose\\na hierarchal Nonnegative Matrix Factorization (NMF)-based method, called\\nConsistent Multiple Nonnegative Matrix Factorization (CMNMF), to factorize\\ngenome-phenome association matrix at two levels of the hierarchical structure\\nin phenotype ontology for mining gene functional modules. CMNMF constrains the\\ngene clusters from the association matrices at two consecutive levels to be\\nconsistent since the genes are annotated with both the child phenotype and the\\nparent phenotype in the consecutive levels. CMNMF also restricts the identified\\nphenotype clusters to be densely connected in the phenotype ontology hierarchy.\\nIn the experiments on mining functionally related genes from mouse phenotype\\nontology and human phenotype ontology, CMNMF effectively improved clustering\\nperformance over the baseline methods. Gene ontology enrichment analysis was\\nalso conducted to reveal interesting gene modules. Conclusions: Utilizing the\\ninformation in the hierarchical structure of phenotype ontology, CMNMF can\\nidentify functional gene modules with more biological significance than the\\nconventional methods. CMNMF could also be a better tool for predicting members\\nof gene pathways and protein-protein interactions. Availability:\\nhttps://github.com/nkiip/CMNMF',\n",
       " 'queries': ['What is Consistent Multiple Nonnegative Matrix Factorization (CMNMF)?',\n",
       "  'How does CMNMF utilize the hierarchical structure of phenotype ontology?',\n",
       "  'What are the advantages of using CMNMF for mining gene functional modules compared to baseline methods?',\n",
       "  'What type of gene modules were revealed by Gene Ontology enrichment analysis in the study?',\n",
       "  'How does CMNMF constrain gene clusters from association matrices at consecutive levels?'],\n",
       " 'hyde_sets': [{'query': 'What is Consistent Multiple Nonnegative Matrix Factorization (CMNMF)?',\n",
       "   'hyde_queries': ['CMNMF is a hierarchical Nonnegative Matrix Factorization (NMF)-based method.',\n",
       "    'CMNMF factorizes genome-phenome association matrix at two levels of phenotype ontology.',\n",
       "    'CMNMF constrains gene clusters from association matrices at consecutive levels to be consistent.',\n",
       "    'CMNMF restricts identified phenotype clusters to be densely connected in the phenotype ontology hierarchy.',\n",
       "    'CMNMF utilizes the hierarchical structure of phenotype ontology to identify functional gene modules.']},\n",
       "  {'query': 'How does CMNMF utilize the hierarchical structure of phenotype ontology?',\n",
       "   'hyde_queries': ['CMNMF factorizes genome-phenome association matrix at two levels of the hierarchical structure in phenotype ontology.',\n",
       "    'CMNMF constrains gene clusters from association matrices at two consecutive levels to be consistent.',\n",
       "    'Genes are annotated with both child phenotype and parent phenotype in consecutive levels of the ontology.',\n",
       "    'CMNMF restricts identified phenotype clusters to be densely connected in the phenotype ontology hierarchy.',\n",
       "    'Hierarchical NMF utilizes information in the hierarchical structure of phenotype ontology.']},\n",
       "  {'query': 'What are the advantages of using CMNMF for mining gene functional modules compared to baseline methods?',\n",
       "   'hyde_queries': ['CMNMF improves clustering performance over baseline methods by utilizing the hierarchical structure of phenotype ontology.',\n",
       "    'CMNMF identifies gene modules with more biological significance than conventional methods.',\n",
       "    'CMNMF constrains gene clusters from association matrices at two consecutive levels to be consistent.',\n",
       "    'CMNMF restricts phenotype clusters to be densely connected in the phenotype ontology hierarchy.',\n",
       "    'CMNMF can be a better tool for predicting members of gene pathways and protein-protein interactions.']},\n",
       "  {'query': 'What type of gene modules were revealed by Gene Ontology enrichment analysis in the study?',\n",
       "   'hyde_queries': ['gene modules related to specific biological processes',\n",
       "    'gene modules involved in protein-protein interactions',\n",
       "    'gene modules associated with specific pathways',\n",
       "    'gene modules showing enrichment for particular GO terms',\n",
       "    'functional gene modules identified in mouse phenotype ontology']},\n",
       "  {'query': 'How does CMNMF constrain gene clusters from association matrices at consecutive levels?',\n",
       "   'hyde_queries': ['CMNMF constrains gene clusters using association matrices at consecutive levels of the phenotype ontology hierarchy.',\n",
       "    'CMNMF enforces consistency between gene clusters derived from association matrices at two consecutive levels.',\n",
       "    'The constraint ensures genes annotated with both child and parent phenotypes are consistently clustered.',\n",
       "    'CMNMF restricts phenotype clusters to be densely connected in the phenotype ontology hierarchy.',\n",
       "    'The method utilizes a hierarchical NMF approach to factorize genome-phenome association matrices at two levels.']}]}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "docs_names = hyde_gen._evaluated_docs()\n",
    "\n",
    "hyde_gen.run(\n",
    "    doc_name=docs_names[2],\n",
    "    thread_id='test_thread'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ac0022",
   "metadata": {},
   "source": [
    "### outputs processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c62d03a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HyDE queries generated for 100 documents.\n"
     ]
    }
   ],
   "source": [
    "hyde_docs = hyde_gen._docs_with_hyde()\n",
    "print(f\"HyDE queries generated for {len(hyde_docs)} documents.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dae8037b",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyde_doc = hyde_gen.get_hyde_file(hyde_docs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6d28e30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'queries_set': [{'query': 'What is complex-valued federated learning?',\n",
       "   'hyde_queries': ['complex-valued federated learning combines federated learning with complex-valued data processing techniques, particularly relevant in fields like MRI.',\n",
       "    'complex-valued federated learning uses differential privacy to protect sensitive data in medical applications like MRI.',\n",
       "    'complex-valued Gaussian mechanism is introduced to address the application of differential privacy to complex-valued data.',\n",
       "    'DP stochastic gradient descent is generalized to complex-valued neural networks in complex-valued federated learning.',\n",
       "    'complex-valued federated learning trains complex-valued neural networks with differential privacy on MRI pulse sequence classification in k-space.']},\n",
       "  {'query': 'How does Differential Privacy enhance federated learning?',\n",
       "   'hyde_queries': ['Differential Privacy (DP) protects sensitive data during federated learning.',\n",
       "    'DP adds noise to gradients, ensuring individual data privacy in federated settings.',\n",
       "    'Federated learning with DP allows training models without direct access to private data.',\n",
       "    'Complex-valued Gaussian mechanism is used to apply DP to complex-valued data in MRI.',\n",
       "    'DP stochastic gradient descent is generalized to complex-valued neural networks for privacy.']},\n",
       "  {'query': 'What are the applications of complex-valued federated learning in MRI?',\n",
       "   'hyde_queries': ['MRI pulse sequence classification in k-space using complex-valued federated learning.',\n",
       "    'Training federated complex-valued neural networks with differential privacy for MRI.',\n",
       "    'Complex-valued signal processing techniques for MRI data acquisition and analysis in federated learning.',\n",
       "    'Application of complex-valued Gaussian mechanism in federated learning for MRI.',\n",
       "    'Privacy-preserving federated learning strategies for MRI data processing.']},\n",
       "  {'query': 'What is the complex-valued Gaussian mechanism?',\n",
       "   'hyde_queries': ['complex-valued Gaussian mechanism for differential privacy',\n",
       "    'Definition of complex-valued Gaussian mechanism in f-DP',\n",
       "    'complex-valued Gaussian mechanism characterized by $(\\\\varepsilon, \\\\delta)$-DP',\n",
       "    'Rényi-DP analysis of complex-valued Gaussian mechanism',\n",
       "    'Application of complex-valued Gaussian mechanism to MRI data']},\n",
       "  {'query': 'How does complex-valued stochastic gradient descent work with DP?',\n",
       "   'hyde_queries': ['complex-valued Gaussian mechanism for DP',\n",
       "    'DP stochastic gradient descent for complex-valued neural networks',\n",
       "    'complex-valued neural network primitives compatible with DP',\n",
       "    'applying DP to complex-valued data',\n",
       "    'federated complex-valued neural networks with DP']}]}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hyde_doc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fe07d0db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[6, 6, 6, 6, 6]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['What is complex-valued federated learning?',\n",
       " 'complex-valued federated learning combines federated learning with complex-valued data processing techniques, particularly relevant in fields like MRI.',\n",
       " 'complex-valued federated learning uses differential privacy to protect sensitive data in medical applications like MRI.',\n",
       " 'complex-valued Gaussian mechanism is introduced to address the application of differential privacy to complex-valued data.',\n",
       " 'DP stochastic gradient descent is generalized to complex-valued neural networks in complex-valued federated learning.',\n",
       " 'complex-valued federated learning trains complex-valued neural networks with differential privacy on MRI pulse sequence classification in k-space.',\n",
       " 'How does Differential Privacy enhance federated learning?',\n",
       " 'Differential Privacy (DP) protects sensitive data during federated learning.',\n",
       " 'DP adds noise to gradients, ensuring individual data privacy in federated settings.',\n",
       " 'Federated learning with DP allows training models without direct access to private data.',\n",
       " 'Complex-valued Gaussian mechanism is used to apply DP to complex-valued data in MRI.',\n",
       " 'DP stochastic gradient descent is generalized to complex-valued neural networks for privacy.',\n",
       " 'What are the applications of complex-valued federated learning in MRI?',\n",
       " 'MRI pulse sequence classification in k-space using complex-valued federated learning.',\n",
       " 'Training federated complex-valued neural networks with differential privacy for MRI.',\n",
       " 'Complex-valued signal processing techniques for MRI data acquisition and analysis in federated learning.',\n",
       " 'Application of complex-valued Gaussian mechanism in federated learning for MRI.',\n",
       " 'Privacy-preserving federated learning strategies for MRI data processing.',\n",
       " 'What is the complex-valued Gaussian mechanism?',\n",
       " 'complex-valued Gaussian mechanism for differential privacy',\n",
       " 'Definition of complex-valued Gaussian mechanism in f-DP',\n",
       " 'complex-valued Gaussian mechanism characterized by $(\\\\varepsilon, \\\\delta)$-DP',\n",
       " 'Rényi-DP analysis of complex-valued Gaussian mechanism',\n",
       " 'Application of complex-valued Gaussian mechanism to MRI data',\n",
       " 'How does complex-valued stochastic gradient descent work with DP?',\n",
       " 'complex-valued Gaussian mechanism for DP',\n",
       " 'DP stochastic gradient descent for complex-valued neural networks',\n",
       " 'complex-valued neural network primitives compatible with DP',\n",
       " 'applying DP to complex-valued data',\n",
       " 'federated complex-valued neural networks with DP']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queries_lengths = [len(hyde_set['hyde_queries'])+1 for hyde_set in hyde_doc['queries_set']]\n",
    "print(queries_lengths)\n",
    "\n",
    "concatenated_queries = [query for queries_set in hyde_doc['queries_set'] for query in [queries_set['query']] + queries_set['hyde_queries']]\n",
    "concatenated_queries"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lang_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
